src/webgpu/shader/validation/functions/restrictions.spec.ts

```diff
@@ -0,0 +1,591 @@
+export const description = `Validation tests for function restrictions`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { keysOf } from '../../../../common/util/data_tables.js';
+import { ShaderValidationTest } from '../shader_validation_test.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+interface VertexPosCase {
+  name: string;
+  value: string;
+  valid: boolean;
+}
+
+const kVertexPosCases: Record<string, VertexPosCase> = {
+  bare_position: { name: `@builtin(position) vec4f`, value: `vec4f()`, valid: true },
+  nested_position: { name: `pos_struct`, value: `pos_struct()`, valid: true },
+  no_bare_position: { name: `vec4f`, value: `vec4f()`, valid: false },
+  no_nested_position: { name: `no_pos_struct`, value: `no_pos_struct()`, valid: false },
+};
+
+g.test('vertex_returns_position')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#function-restriction')
+  .desc(`Test that a vertex shader should return position`)
+  .params(u => u.combine('case', keysOf(kVertexPosCases)))
+  .fn(t => {
+    const testcase = kVertexPosCases[t.params.case];
+    const code = `
+struct pos_struct {
+  @builtin(position) pos : vec4f
+}
+
+struct no_pos_struct {
+  @location(0) x : vec4f
+}
+
+@vertex
+fn main() -> ${testcase.name} {
+  return ${testcase.value};
+}`;
+
+    t.expectCompileResult(testcase.valid, code);
+  });
+
+g.test('entry_point_call_target')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#function-restriction')
+  .desc(`Test that an entry point cannot be the target of a function call`)
+  .params(u =>
+    u.combine('stage', ['@fragment', '@vertex', '@compute @workgroup_size(1,1,1)'] as const)
+  )
+  .fn(t => {
+    let ret_attr = '';
+    if (t.params.stage === '@vertex') {
+      ret_attr = '@builtin(position)';
+    }
+    const with_entry_point = `
+${t.params.stage}
+fn bar() -> ${ret_attr} vec4f {
+  return vec4f();
+}
+
+fn foo() {
+  let tmp = bar();
+}
+`;
+    const without_entry_point = `
```

Comment:
Would this be better making the with/without entry point a param that is passed in, then the same shader ends up being used for both the pass/fail cases?

Reply:
sure, I'll make that change.